text/3374-cargo-target-features.md

```diff
@@ -0,0 +1,177 @@
+
+# Summary
+[summary]: #summary
+
+This adds a new `enable-features` field to [Cargo Targets](https://doc.rust-lang.org/cargo/reference/cargo-targets.html#configuring-a-target) which forces a [feature](https://doc.rust-lang.org/cargo/reference/features.html) to be enabled if the target is being built.
+
+# Motivation
+[motivation]: #motivation
+
+There are several situations where a non-library target must have a particular feature enabled for it to work correctly.
+Although it is possible to manually enable these features on the command-line, this can make it awkward to use.
+This RFC adds a mechanism to make it easier to work with these targets.
+
+Some example use cases are:
+
+1. Binaries that need additional dependencies like command-line processing libraries, or logging functionality.
+2. Examples illustrating how to use a library with a specific feature enabled.
+3. Benchmarks which require a separate benchmarking library, but you don't want to pay the cost of building that library when not running benchmarks.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The `enable-features` field can be added to a target in `Cargo.toml` to specify features or dependencies that will be automatically enabled whenever the target is being built.
+For example:
+
+```toml
+[package]
+name = "myproject"
+version = "0.1.0"
+edition = "2021"
+
+[[example]]
+name = "fetch-http"
+enable-features = ["networking"]
+
+[[bin]]
+name = "myproject"
+enable-features = ["dep:clap"]
+
+[dependencies]
+clap = { version = "4.0.26", optional = true }
+
+[features]
+networking = []
+```
+
+When using `myproject` as a library dependency, by default the `networking` feature and `clap` will not be enabled.
+When building the binary as in `cargo install myproject` or `cargo build`, the `clap` dependency will automatically be enabled allowing the binary to be built correctly.
+Similarly, `cargo build --example fetch-http` will enable the `networking` feature so that the example can be built.
+
+This field can be specified for any Cargo Target except the `[lib]` target.
+
+> **Note**: Because features and dependencies are package-wide, using `enable-features` does not narrow the scope of the features or dependencies to the specific target.
+> The features and dependencies will be enabled for all targets.
+
+# Implementation Details
+
+## Feature resolver and selection
+
+Before doing dependency and feature resolution, Cargo will need to determine which features are being force-enabled by the targets selected on the command-line.
+These additional features will be combined with any features selected on the command-line before calling the resolver.
+
+Unfortunately this selection of targets is quite complex.
+This may require some potentially invasive refactoring, as Cargo currently chooses the targets to select after resolution is done (including a separate phase for implicit binaries used for integration tests).
```

Comment:
Quoting your question from https://github.com/rust-lang/rfcs/pull/3020#issuecomment-862381779:

> I think the interaction with required-features could use some clarification in this scenario:
> 
> ```toml
> [[bin]]
> name = "foo"
> lib-features = ["a"]
> 
> [[bin]]
> name = "bar"
> required-features = ["a"]
> lib-features = ["b"]
> 
> [[bin]]
> name = "baz"
> required-features = ["b"]
> ```
> 
> Here, building with no feature flags, I would assume the resolver would see `foo` enables `a`. Once `a` is enabled, that means `bar`'s required-features are now satisfied, so it can be built. That in turn enables `b` which in turn means `baz`'s required-features are now satisfied, so it can be built. That is what I mean by an iterative approach.
> 
> If it doesn't work like that, then there is some alternative that needs to be clarified. I'd be concerned if features get enabled, but they don't allow `required-features` to be satisfied, since that could be confusing. Although it is not really clear if required-features will be used much, so it might not be too important, it is good to clarify how this works.

How does the feature resolver solve this problem?

Reply:
Thanks for bringing that up! I had forgotten about it, and have added it to the RFC. It should likely iterate until no new requirements can be satisfied.